vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+                                                               RRNodeId(sink),
+                                                               sink == track_id ? 1 : 0);
+            }
+        }
+    }
+
+    void pack_chans_edges() {
+        for (auto& i : virtual_shorts_) {
+            location l1, l2;
+            e_rr_type t1, t2;
+            int virtual_idx1, idx1, virtual_idx2, idx2;
+            std::tie(l1, t1, virtual_idx1, l2, t2, virtual_idx2) = i;
+            idx1 = virtual_beg_to_real_[std::make_tuple(l1, t1, virtual_idx1)];
+            idx2 = virtual_beg_to_real_[std::make_tuple(l2, t2, virtual_idx2)];
+
+            std::tuple<location, e_rr_type, int> key1(l1, t1, idx1), key2(l2, t2, idx2);
+
+            int src, sink;
+            src = loc_type_idx_to_rr_idx_[key1];
+            sink = loc_type_idx_to_rr_idx_[key2];
+
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(src), RRNodeId(sink), 0);
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(sink), RRNodeId(src), 0);
+        }
+    }
+
+    char* int_to_string(char* buff, int value) {
+        if (value < 10) {
+            return &(*buff = '0' + value) + 1;
+        } else {
+            return &(*int_to_string(buff, value / 10) = '0' + value % 10) + 1;
+        }
+    }
+
+    void pack_pips() {
+        for (auto& i : pips_) {
+            int node1, node2;
+            std::tie(node1, node2) = i.first;
+            int sw_id, tile_id;
+            std::tuple<int, int, int, bool> metadata;
+            std::tie(sw_id, tile_id, metadata) = i.second;
+            VTR_ASSERT(sw_id > 1);
+            int name, wire0, wire1;
+            bool forward;
+            std::tie(name, wire0, wire1, forward) = metadata;
+            location loc = tile_to_loc_[tile_id];
+            auto virtual_key1 = virtual_redirect_[std::make_tuple(node1, loc)];
+            auto virtual_key2 = virtual_redirect_[std::make_tuple(node2, loc)];
+
+            auto key1 = std::make_tuple(std::get<0>(virtual_key1),
+                                        std::get<1>(virtual_key1),
+                                        virtual_beg_to_real_[virtual_key1]);
+            auto key2 = std::make_tuple(std::get<0>(virtual_key2),
+                                        std::get<1>(virtual_key2),
+                                        virtual_beg_to_real_[virtual_key2]);
+
+            int src, sink;
+            src = loc_type_idx_to_rr_idx_[key1];
+            sink = loc_type_idx_to_rr_idx_[key2];
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(src), RRNodeId(sink), sw_id);
+
+            char metadata_[100];
+            char* temp = int_to_string(metadata_, name);
+            *temp++ = ',';
+            temp = int_to_string(temp, wire0);
+            *temp++ = ',';
+            temp = int_to_string(temp, wire1);
+            *temp++ = ',';
+            temp = int_to_string(temp, forward ? 1 : 0);
+            *temp++ = 0;
```

Comment:
It would be better to use c++ standards here and maybe have a string stream to build up the metadata string

Reply:
I tried to use c++ standard, but it was too slow. This change alone speed up RR generation by 3 seconds on nexus 17 device

Reply:
Ok. I believe than that the `int_to_string` function should end up either in the `vtrutil` libraries or at least external to the RR graph reader as it does not concern the RR graph per se

Reply:
Moved to `fpga_interchange_arch_utils`